src/otlp.zig

```diff
@@ -0,0 +1,744 @@
+    const Self = @This();
+
+    allocator: std.mem.Allocator,
+    client: *http.Client,
+
+    max_retries: u32,
+    base_delay_ms: u64,
+    max_delay_ms: u64,
+
+    queue: std.PriorityQueue(FetchRequest, void, compareFetchRequest),
+    thread: ?std.Thread,
+    mutex: std.Thread.Mutex,
+    condition: std.Thread.Condition,
+    running: std.atomic.Value(bool),
+
+    // Wrapper for FetchOptions with retry metadata
+    const FetchRequest = struct {
+        options: http.Client.FetchOptions,
+        attempts_counter: u32,
+        next_attempt_time_ms: i64, // Timestamp (ms) for next retry
+    };
+
+    // Compare function for PriorityQueue (earlier next_attempt first)
+    fn compareFetchRequest(_: void, a: FetchRequest, b: FetchRequest) std.math.Order {
+        return std.math.order(a.next_attempt_time_ms, b.next_attempt_time_ms);
+    }
+
+    fn init(allocator: std.mem.Allocator, client: *http.Client, config: struct {
+        max_retries: u32 = 50,
+        base_delay_ms: u64 = 100,
+        max_delay_ms: u64 = 60000,
+    }) !*Self {
+        const s = try allocator.create(Self);
+        s.* = Self{
+            .allocator = allocator,
+            .client = client,
+            .max_retries = config.max_retries,
+            .base_delay_ms = config.base_delay_ms,
+            .max_delay_ms = config.max_delay_ms,
+            .queue = std.PriorityQueue(FetchRequest, void, compareFetchRequest).init(allocator, {}),
+            .mutex = std.Thread.Mutex{},
+            .thread = null,
+            .condition = std.Thread.Condition{},
+            .running = std.atomic.Value(bool).init(true),
+        };
+
+        // Start the background thread
+        s.thread = try std.Thread.spawn(.{}, retryLoop, .{s});
+        return s;
+    }
+
+    fn deinit(self: *Self) void {
+        // Signal the thread to stop
+        self.running.store(false, .release);
+        // Wake the thread if it's waiting
+        self.condition.signal();
+
+        // Wait for the thread to finish
+        if (self.thread) |t| {
+            t.join();
+        }
+
+        // Clean up queue
+        self.mutex.lock();
+        while (self.queue.removeOrNull()) |req| {
+            self.freeFetchRequest(&req);
+        }
+        self.queue.deinit();
+        self.mutex.unlock();
+
+        self.allocator.destroy(self);
+    }
+
+    fn enqueue(self: *Self, options: http.Client.FetchOptions, attempt: u32) !void {
+        const req = FetchRequest{
+            .options = try self.cloneFetchOptions(options),
+            .attempts_counter = attempt,
+            .next_attempt_time_ms = std.time.milliTimestamp() + calculateDelay(self.base_delay_ms, self.max_delay_ms, attempt),
+        };
+
+        self.mutex.lock();
+        defer self.mutex.unlock();
+        try self.queue.add(req);
+        self.condition.signal();
+    }
+
+    fn retryLoop(self: *Self) void {
+        while (self.running.load(.acquire)) {
+            var req: FetchRequest = undefined;
```

Comment:
There's a deadlock here: since we `.acquire` the atomic in each loop iteration, but never release it, it only does the first retry and blocks forever.